compiler/rustc_smir/src/rustc_internal/mod.rs

```diff
@@ -197,3 +185,26 @@ where
         })
     }
 }
+
+/// Simmilar to rustc's `FxIndexMap`, `IndexMap` with extra
+/// safety features added.
+pub struct IndexMap<K, V> {
+    index_map: fx::FxIndexMap<K, V>,
+}
+
+impl<K: PartialEq + Hash + Eq, V: Copy + Debug + PartialEq + IndexToVal> IndexMap<K, V> {
+    /// Because we are using values as indexes, this first get's the
+    /// key using index of the value provided, then asserts that value
+    /// we got from IndexMap is equal to one we provided.
+    pub fn get_index_and_assert(&self, index: usize, value: V) -> &K {
```

Comment:
Once the the other direction also exists on the trait, you don't need to pass the `index` manually anymore and can get it from the `V`

Reply:
Also the `assert` part is an internal detail, can leave it out of the name. Could even implement `Index` at that point instead of having an inherent method.